pkg/corerp/api/v20231001preview/environment_conversion.go

```diff
@@ -408,35 +420,118 @@ func fromRecipePropertiesClassificationDatamodel(e datamodel.EnvironmentRecipePr
 	return nil
 }
 
-func toRecipeConfigTerraformProvidersDatamodel(config *RecipeConfigProperties) map[string][]datamodel.ProviderConfigProperties {
+func toRecipeConfigTerraformProvidersDatamodel(config *RecipeConfigProperties) (map[string][]datamodel.ProviderConfigProperties, error) {
 	if config.Terraform == nil || config.Terraform.Providers == nil {
-		return nil
+		return nil, nil
 	}
 
 	dm := map[string][]datamodel.ProviderConfigProperties{}
 	for k, v := range config.Terraform.Providers {
 		dm[k] = []datamodel.ProviderConfigProperties{}
 
 		for _, providerAdditionalProperties := range v {
+			var addnlProperties map[string]any
+			var convertedSecrets map[string]datamodel.SecretReference
+
+			convertedSecrets = getSecretsFromProviderProperties(providerAdditionalProperties.Secrets)
+			addnlProperties = providerAdditionalProperties.AdditionalProperties
+
+			// Account for possibility of nested additional properties during unmarshalling
+			// https://json-schema.org/understanding-json-schema/reference/object#additionalproperties
+			if tmpAddnlProperties, ok := addnlProperties["additionalProperties"]; ok && addnlProperties["additionalProperties"] != nil {
+				if tmpMap, ok := tmpAddnlProperties.(map[string]any); ok {
+					addnlProperties = tmpMap
+				} else {
+					return nil, v1.ErrInvalidModelConversion
+				}
+			}
+
+			// check if additional properties contain Secrets
+			// get AdditionalProperties from providerAdditionalProperties excluding the 'secrets' key.
+			for key, value := range addnlProperties {
+				if key == "secrets" {
+					if tmpSecret, ok := value.(map[string]*SecretReference); !ok {
+						return nil, v1.ErrInvalidModelConversion
+					} else {
+						propSecrets := getSecretsFromProviderProperties(tmpSecret)
+
+						// Add the secrets to the convertedSecrets map
+						for secretKey, secretValue := range propSecrets {
+							if convertedSecrets == nil {
+								convertedSecrets = map[string]datamodel.SecretReference{}
+							}
+							convertedSecrets[secretKey] = secretValue
+						}
+					}
+
+					// remove the 'secrets' key from the additional properties as we've added all
+					// secrets to convertedSecrets map
+					delete(addnlProperties, "secrets")
+				}
+			}
+
 			dm[k] = append(dm[k], datamodel.ProviderConfigProperties{
-				AdditionalProperties: providerAdditionalProperties,
+				AdditionalProperties: addnlProperties,
+				Secrets:              convertedSecrets,
 			})
 		}
 	}
 
-	return dm
+	return dm, nil
+}
+
+func getSecretsFromProviderProperties(providerProperties map[string]*SecretReference) map[string]datamodel.SecretReference {
```

Comment:
We have a specific convention for function names for to/from conversions between internal and external data models, can you update this function name to follow that?

Reply:
updated.